### PR TITLE
Constant propagation update

### PR DESCRIPTION
**Pull request by Gustavo2622**
Makes it so that constant propagation doesn't auto stop at first write and continues as far as possible. If # of lines parameter is given, tries to propagate to given number and fails if not possible.

### CODE REVIEW

File: src/phl/ecPhlCodeTx.ml [hunk 1]
```diff
@@ -136,68 +137,140 @@ let t_set_r side cpos (fresh, id) e tc =
...
+    (insts: instr list) 
+    : instr list =
+    let (asgn, tl) =
+      match insts with
+      | ({ i_node = Sasgn (lv, e) }) :: tl -> begin
+        let asgn =
+          match lv with
+          | LvVar (x, ty) -> 
+            (* check if first iteration or if variable is being propagated *)
+            let () = assert (asg = PV.empty || PV.mem_pv env x asg) in
```
Comment (strub) [inline src/phl/ecPhlCodeTx.ml:157]: Prefer `s1; s2` in place of `let () = s1 in s2`

File: src/phl/ecPhlCodeTx.ml [hunk 2]
```diff
@@ -136,68 +137,140 @@ let t_set_r side cpos (fresh, id) e tc =
...
+            [(x, ty, e)]
+          | LvTuple xs -> begin
+              match e.e_node with
+              | Etuple es -> 
+                (* check if first iteration or if all the variables in the tuple are being propagated *)
+                let () = assert (asg = PV.empty || 
+                    List.for_all (fun x -> PV.mem_pv env x asg) (List.fst xs)) in
```
Comment (strub) [inline src/phl/ecPhlCodeTx.ml:164]: Bis

File: src/phl/ecPhlCodeTx.ml [hunk 3]
```diff
@@ -136,68 +137,140 @@ let t_set_r side cpos (fresh, id) e tc =
...
-  in
+    (* Check that we are assigning to local variables only *)
+    List.iter
+      (fun (x, _, _) ->
+        if is_glob x then
+          tc_error pf "left-values must be local variables")
+      asgn;
+
+    (* Check that we are assigning an expression that we can evaluate to a constant  *)
+    let expr_is_closed (e: expr) : bool = e_fv e = Mid.empty && e_read env e = PV.empty in
```
Comment (strub) [inline src/phl/ecPhlCodeTx.ml:184]: Use `Mid.is_empty` (I am not sure there is a guarantee that all empty maps are structurally equal to `Mid.empty`)

Same for `PV.empty` (add a `PV.is_empty` function and use it)

File: src/phl/ecPhlCodeTx.ml [hunk 4]
```diff
@@ -136,68 +137,140 @@ let t_set_r side cpos (fresh, id) e tc =
...
+    
+    let do_subst = Mpv.isubst env subst in
+    let do_substs = Mpv.issubst env subst in
+
+    (* Rebuild assignments *)
+    let asgn_instrs_from_pv_and_subst (asg) (subst) : instr = 
+      let asgns = fst (PV.elements asg) in
+      match asgns with 
+      | [] -> assert false (* meaning we dont propagate, should never happen *)
+      | (pv, ty)::[] -> i_asgn (LvVar (pv, ty), e_var pv ty) |> (Mpv.isubst env subst)
```
Comment (strub) [inline src/phl/ecPhlCodeTx.ml:213]: Use `[x]`, not `x :: []`.

Coding style: we tend to put spaces around binary operators.

**Review by strub — CHANGES_REQUESTED** [review]

File: src/phl/ecPhlCodeTx.ml (hunk 3 above)
Comment (Gustavo2622) [reply src/phl/ecPhlCodeTx.ml:184]: There is already a PV.is_empty function

**Pull request closed without merging**